apps/ex_wire/lib/ex_wire/tcp/listener.ex

```diff
@@ -0,0 +1,46 @@
+defmodule ExWire.TCP.Listener do
+  @moduledoc """
+  Module responsible for opening a tcp listen socket to listen for incoming
+  connections. Once a connection is accepted, it will hand off control of the
+  accepted socket connection to a separate process.
+  """
+
+  use GenServer
+
+  alias ExWire.TCP
+
+  def start_link(args) do
+    port = Keyword.fetch!(args, :port)
+    name = Keyword.fetch!(args, :name)
+
+    GenServer.start_link(__MODULE__, port, name: name)
+  end
+
+  def init(port) do
+    {:ok, listen_socket} = TCP.listen(port)
+
+    accept_tcp_connection()
+
+    {:ok, %{listen_socket: listen_socket}}
+  end
+
+  @doc """
+  Accepts a connection, and gives control of the connection to a separate process
+  that will henceforth handle that tcp connection.
+  """
+  def handle_cast(:accept_tcp_connection, state = %{listen_socket: listen_socket}) do
+    {:ok, socket} = TCP.accept_connection(listen_socket)
+    {:ok, pid} = TCP.InboundConnectionsSupervisor.new_connection_handler()
+
+    TCP.hand_over_control(socket, pid)
```

Comment:
status of these two operations are not  being checked. they can fail. is it ok?

Reply:
That's a really good question. I'll do some due diligence on that. My original thought was that we don't care if these fail because that'll just drop the connection in the other process. The `listener` process should not fail because of bad response. But... now that I'm looking at this, if the hand over fails, then I want to make sure this process would not get the messages in its mailbox, so I'll do a little digging.

Reply:
We decided to make sure the hand-over of control is successful before accepting messages. Changes in 8fde0645d358917ba545d2a6571b3a28c3abc1b1